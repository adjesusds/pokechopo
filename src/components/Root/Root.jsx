// External Packages
import { node } from 'prop-types'
import React from 'react'

// Internal Packages
import StyledRoot from './styled/Root'

/**
 * Creates a new Root component.
 * @class
 */
const Root = ({ children }) => <StyledRoot>{children}</StyledRoot>

/** @override */
Root.propTypes = {
  children: node.isRequired,
}

export default Root
