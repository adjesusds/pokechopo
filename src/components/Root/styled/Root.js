// External Packages
import styled from 'styled-components'

/**
 * @constant {Element} style The style of the component.
 */
const style = styled.div`
  align-items: center;
  display: flex;
  flex-direction: column;
`

export default style
