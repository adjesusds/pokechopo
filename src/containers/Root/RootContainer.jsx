// External Packages
import { bool, func, number, shape, string } from 'prop-types'
import React, { Fragment } from 'react'
import { connect } from 'react-redux'

// Internal Packages
import { fetchPokemon } from 'actions/pokemon'
import { Button, Copyright, Divider, Link, Loading, Message } from 'components'
import constants from 'config/constants'

/**
 * Creates a new Root container.
 * @class
 */
const RootContainer = ({ data, errorMessage, isFetching, onDiscoverPokemonClick }) => (
  <Fragment>
    <Button onClick={onDiscoverPokemonClick} disabled={isFetching}>
      Discover a Pok√©mon
    </Button>
    <Divider />
    {!data ? (
      <Fragment>
        <Message variant="warning">{!isFetching ? 'Click it.' : <Loading message="Searching" />}</Message>
        {!errorMessage ? (
          <Message>
            Gotta catch &#39;em all!<Message variant="info">&trade;</Message>
          </Message>
        ) : (
          <Message variant="danger">{errorMessage}</Message>
        )}
      </Fragment>
    ) : (
      <Fragment>
        <Message>
          <Link href={`https://www.pokemon.com/us/pokedex/${data.name}/`} target="blank">
            {`${data.name[0].toUpperCase()}${data.name.slice(1)}`}
          </Link>
          <Message variant="warning">
            {` #${String(data.id).padStart(String(constants.POKEMON_MAX_NUMBER).length, '0')}`}
          </Message>
        </Message>
        <Message variant="success">Look at what you got!</Message>
      </Fragment>
    )}
    <Copyright>
      &copy; {new Date().getFullYear()}{' '}
      <Link href="https://www.linkedin.com/in/relztic/" target="blank">
        Relztic
      </Link>{' '}
      üê¥
    </Copyright>
  </Fragment>
)

/** @override */
RootContainer.propTypes = {
  data: shape({
    id: number.isRequired,
    name: string.isRequired,
  }),
  errorMessage: string.isRequired,
  isFetching: bool.isRequired,
  onDiscoverPokemonClick: func.isRequired,
}

/** @override */
RootContainer.defaultProps = {
  data: null,
}

/** @override */
const mapStateToProps = state => state.pokemon

/** @override */
const mapDispatchToProps = dispatch => ({
  onDiscoverPokemonClick: event => {
    event.preventDefault()

    const pokemonNumber = Math.floor(Math.random() * constants.POKEMON_MAX_NUMBER) + constants.POKEMON_MIN_NUMBER

    dispatch(fetchPokemon(pokemonNumber))
  },
})

/**
 * @constant {Element} ConnectedRootContainer The connected Root container, Redux.
 */
const ConnectedRootContainer = connect(mapStateToProps, mapDispatchToProps)(RootContainer)

export default ConnectedRootContainer

export { RootContainer as PlainRootContainer, mapDispatchToProps, mapStateToProps }
