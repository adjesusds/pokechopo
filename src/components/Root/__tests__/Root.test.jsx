// External Packages
import { shallow } from 'enzyme'
import React from 'react'

// Internal Packages
import Root from '../Root'

describe('<Root />', () => {
  it('should render the component on valid props', () => {
    const component = shallow(<Root>Hello, World!</Root>)

    expect(component).toMatchSnapshot()
  })
})
