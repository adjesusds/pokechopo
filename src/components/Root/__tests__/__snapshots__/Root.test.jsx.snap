// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Root /> should render the component on valid props 1`] = `
<styled.div>
  Hello, World!
</styled.div>
`;
